pkg/controller/volume/attach_detach_controller.go

```diff
@@ -106,30 +240,364 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+			key,
+			err)
+	}
+
+	pvc, ok := pvcObj.(*api.PersistentVolumeClaim)
+	if ok || pvc == nil {
+		return "", "", fmt.Errorf(
+			"failed to cast %q object %#v to PersistentVolumeClaim",
+			key,
+			pvcObj)
+	}
+
+	if pvc.Status.Phase != api.ClaimBound || pvc.Spec.VolumeName == "" {
+		return "", "", fmt.Errorf(
+			"PVC %q has non-bound phase (%q) or empty pvc.Spec.VolumeName (%q)",
+			key,
+			pvc.Status.Phase,
+			pvc.Spec.VolumeName)
+	}
+
+	return pvc.Spec.VolumeName, pvc.UID, nil
+}
+
+// getPVSpecFromCache fetches the PV object with the given name from the shared
+// internal PV store and returns a volume.Spec representing it.
+// This method returns an error if a PV object does not exist in the cache with
+// the given name.
+// This method deep copies the PV object so the caller may use the returned
+// volume.Spec object without worrying about it mutating unexpectedly.
+func (adc *attachDetachController) getPVSpecFromCache(
+	name string,
+	pvcReadOnly bool,
+	expectedClaimUID types.UID) (*volume.Spec, error) {
+	pvObj, exists, err := adc.internalPVInformer.GetStore().Get(name)
+	if pvObj == nil || !exists || err != nil {
+		return nil, fmt.Errorf(
+			"failed to find PV %q in PVInformer cache. %v", name, err)
+	}
+
+	pv, ok := pvObj.(*api.PersistentVolume)
+	if ok || pv == nil {
+		return nil, fmt.Errorf(
+			"failed to cast %q object %#v to PersistentVolume", name, pvObj)
+	}
+
+	if pv.Spec.ClaimRef == nil {
+		return nil, fmt.Errorf(
+			"found PV object %q but it has a nil pv.Spec.ClaimRef indicating it is not yet bound to the claim",
+			name)
+	}
+
+	if pv.Spec.ClaimRef.UID != expectedClaimUID {
+		return nil, fmt.Errorf(
+			"found PV object %q but its pv.Spec.ClaimRef.UID (%q) does not point to claim.UID (%q)",
+			name,
+			pv.Spec.ClaimRef.UID,
+			expectedClaimUID)
+	}
+
+	// Do not return the object from the informer, since the store is shared it
+	// may be mutated by another consumer.
+	clonedPVObj, err := api.Scheme.DeepCopy(pv)
+	if err != nil || clonedPVObj == nil {
+		return nil, fmt.Errorf("failed to deep copy %q PV object", name)
+	}
+
+	clonedPV, ok := clonedPVObj.(*api.PersistentVolume)
+	if ok || clonedPV == nil {
+		return nil, fmt.Errorf(
+			"failed to cast %q clonedPV %#v to PersistentVolume", name, pvObj)
+	}
+
+	return volume.NewSpecFromPersistentVolume(clonedPV, pvcReadOnly), nil
+}
+
+// processSafeToDetachAnnotations processes the "safe to detach" annotations for
+// the given node. It makes calls to delete any annotations refering to volumes
+// it is not aware of. For volumes it is aware of, it marks them safe to detach
+// in the "actual state of world" data structure.
+func (adc *attachDetachController) processSafeToDetachAnnotations(
+	nodeName string, annotations map[string]string) {
+	var annotationsToRemove []string
+	for annotation := range annotations {
+		// Check annotations for "safe to detach" volumes
+		annotation = strings.ToLower(annotation)
```

Comment:
As discussed above, safeToDetach annotation stuff will be removed in follow up PR in favor of new Node Status field.